include/zephyr/pm/pm.h

```diff
@@ -53,16 +53,25 @@ extern "C" {
  */
 struct pm_notifier {
 	sys_snode_t _node;
+
 	/**
-	 * Application defined function for doing any target specific operations
+	 * Value indicating whether the callback should trigger on entry or
+	 * exit of a power state.
+	 */
+	enum pm_state_direction direction;
+
+	/**
+	 * Application or driver defined function for doing any target specific operations
 	 * for power state entry.
 	 */
-	void (*state_entry)(enum pm_state state);
+	void (*callback)(void *arg1, void *arg2, void *arg3);
+
 	/**
-	 * Application defined function for doing any target specific operations
-	 * for power state exit.
+	 * Optional callback arguments
 	 */
-	void (*state_exit)(enum pm_state state);
+	void *arg1;
+	void *arg2;
+	void *arg3;
```

Comment:
note here: I'd not repeat the ugly triple void * found in threads. I think a single `void *ctx` should be enough, or `const struct device *dev` if this API is expected to always be used in the context of devices.

Reply:
Good point, a single (const) void pointer would probably be sufficient.

I think the `pm_notifier` design was initially built to target application code, so I didn't want to directly add a device instance pointer, but leave it open for potential use by app code as well as drivers.